### PR TITLE
Implement Tape to record API calls

### PR DESCRIPTION
**Pull request by msz-rai**

### CODE REVIEW

File: src/api/api.cpp [hunk 1]
```diff
@@ -141,34 +151,53 @@ rgl_get_version_info(int32_t* out_major, int32_t* out_minor, int32_t* out_patch)
...
+	return status;
+}
+
+void TapePlay::tape_get_version_info(const YAML::Node& yamlNode)
+{
+	int32_t out_major, out_minor, out_patch;
+	rgl_get_version_info(&out_major, &out_minor, &out_patch);
+	if (out_major != yamlNode["0"].as<int32_t>()) RGL_WARN("tape_get_version_info: out_major mismatch");
+	if (out_minor != yamlNode["1"].as<int32_t>()) RGL_WARN("tape_get_version_info: out_minor mismatch");
+	if (out_patch != yamlNode["2"].as<int32_t>()) RGL_WARN("tape_get_version_info: out_patch mismatch");
```
Comment (prybicki) [inline src/api/api.cpp:177]: I'd go with braces ;)

File: src/api/api.cpp [hunk 2]
```diff
@@ -141,34 +151,53 @@ rgl_get_version_info(int32_t* out_major, int32_t* out_minor, int32_t* out_patch)
...
 		auto logFilePath = log_file_path == nullptr ? std::nullopt : std::optional(log_file_path);
 		Logger::instance().configure(log_level, logFilePath, use_stdout);
 	});
+	TAPE_HOOK(log_level, log_file_path, use_stdout);
+	return status;
+}
+
+void TapePlay::tape_configure_logging(const YAML::Node& yamlNode)
+{
+	rgl_configure_logging((rgl_log_level_t)yamlNode["0"].as<int>(), yamlNode["1"].as<std::string>().c_str(), yamlNode["2"].as<bool>());
```
Comment (prybicki) [inline src/api/api.cpp:196]:
```suggestion
	rgl_configure_logging(
	    (rgl_log_level_t) yamlNode["0"].as<int>(),
	    yamlNode["1"].as<std::string>().c_str(),
	    yamlNode["2"].as<bool>()
	);
```

File: src/api/api.cpp [hunk 3]
```diff
@@ -213,13 +246,23 @@ rgl_cleanup(void)
...
 
 RGL_API rgl_status_t
-rgl_mesh_create(rgl_mesh_t *out_mesh, const rgl_vec3f *vertices, int32_t vertex_count, const rgl_vec3i *indices, int32_t index_count)
+rgl_mesh_create(rgl_mesh_t* out_mesh, const rgl_vec3f* vertices, int32_t vertex_count, const rgl_vec3i* indices, int32_t index_count)
 {
-	return rglSafeCall([&]() {
-        RGL_DEBUG("rgl_mesh_create(out_mesh={}, vertices={}, indices={})",
+	auto status = rglSafeCall([&]() {
+		RGL_DEBUG("rgl_mesh_create(out_mesh={}, vertices={}, indices={})",
 			  (void*) out_mesh, repr(vertices, vertex_count), repr(indices, index_count, 1));
```
Comment (prybicki) [inline src/api/api.cpp:266]:
```suggestion
	auto status = rglSafeCall([&]() {
		RGL_DEBUG("rgl_mesh_create(out_mesh={}, vertices={}, indices={})",
		         (void*) out_mesh, repr(vertices, vertex_count), repr(indices, index_count, 1));
```

File: src/api/api.cpp [hunk 4]
```diff
@@ -231,35 +274,66 @@ rgl_mesh_create(rgl_mesh_t *out_mesh, const rgl_vec3f *vertices, int32_t vertex_
...
+
+void TapePlay::tape_mesh_create(const YAML::Node& yamlNode)
+{
+	rgl_mesh_t mesh = nullptr;
+	rgl_mesh_create(&mesh,
+		reinterpret_cast<const rgl_vec3f*>(fileMmap + yamlNode["1"].as<size_t>()),
+		yamlNode["2"].as<int>(),
+		reinterpret_cast<const rgl_vec3i*>(fileMmap + yamlNode["3"].as<size_t>()),
+		yamlNode["4"].as<int>());
+	tapeMeshes.insert(std::make_pair(yamlNode["0"].as<size_t>(), mesh));
```
Comment (prybicki) [inline src/api/api.cpp:289]: Is it possible to address `yamlNode` using integer instead of string? `yamlNode[0]` looks better.

**Review by prybicki — COMMENTED** [review]
Great job! :tada: 
Left some suggestions and comments.
Good to see it with tests and visualization extensions! :+1:

File: src/Tape.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+// limitations under the License.
+
+#include "Tape.hpp"
+
+std::optional<TapeRecord> tapeRecord;
+
+TapeRecord::TapeRecord(const char* path)
+{
+	std::string pathYaml = std::string(path) + YAML_EXTENSION;
+	std::string pathBin = std::string(path) + BIN_EXTENSION;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.cpp:22]: Thanks for the suggestion

File: src/Tape.hpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+#define PROT_READ 1
+#define MAP_PRIVATE 1
+#define MAP_FAILED nullptr
+static void munmap(void* addr, size_t length) { }
+static void* mmap(void* start, size_t length, int prot, int flags, int fd, size_t offset) { return nullptr; }
+#else
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif // _WIN32
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.hpp:29]: Looks better, thanks

File: src/Tape.hpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	intptr_t valueToYaml(rgl_entity_t value) { return (intptr_t)value; }
+	intptr_t valueToYaml(rgl_entity_t* value) { return (intptr_t)*value; }
+
+	intptr_t valueToYaml(rgl_mesh_t value) { return (intptr_t)value; }
+	intptr_t valueToYaml(rgl_mesh_t* value) { return (intptr_t)*value; }
+
+	intptr_t valueToYaml(rgl_scene_t value) { return (intptr_t)value; }
+	intptr_t valueToYaml(rgl_scene_t* value) { return (intptr_t)*value; }
+
+	intptr_t valueToYaml(void* value) { return (intptr_t)value; }
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.hpp:119]: Good catch, thanks

File: src/Tape.hpp [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+		}
+		yamlRecording.push_back(apiCallNode);
+	}
+};
+
+class TapePlay
+{
+	YAML::Node yamlRoot;
+	uint8_t* fileMmap{};
+	size_t mmapSize{};
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.hpp:152]: I would leave it as it is. The class could be helpful when we decide to implement more features to tape in the future

File: src/api/api.cpp [hunk 9]
```diff
@@ -141,34 +151,53 @@ rgl_get_version_info(int32_t* out_major, int32_t* out_minor, int32_t* out_patch)
 	// 0.9.1: optimized rgl_mesh_update_vertices
 	// 0.9.2: remove external dependency (gdt)
 	// 0.9.3: add API unit tests with fixes, in particular: optix logging
-	// 0.10.0: entities can now share meshes
+	// 0.10.0: tapeEntities can now share tapeMeshes
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/api/api.cpp:154]: Find&Replace stuff. Fixed

File: src/api/api.cpp (hunk 4 above)
Comment (msz-rai) [reply src/api/api.cpp:289]: Yes, it is possible. Changed

File: src/api/api.cpp [hunk 10]
```diff
@@ -518,5 +814,75 @@ rgl_node_points_visualize(rgl_node_t* node, const char* window_name, int32_t win
...
+	#endif
+	return rglSafeCall([&]() {
+		CHECK_ARG(path != nullptr);
+		RGL_DEBUG("rgl_tape_record_begin(path={})", path);
+		if (tapeRecord.has_value()) {
+			throw RecordError("rgl_tape_record_begin: recording already active");
+		} else {
+			tapeRecord.emplace(path);
+		}
+	});
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/api/api.cpp]: I can't see any errors

File: src/Tape.hpp [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+#include <Logger.hpp>
+#include <RGLExceptions.hpp>
+#include <rgl/api/core.h>
+
+#define BIN_EXTENSION ".bin"
+#define YAML_EXTENSION ".yaml"
+#define RGL_VERSION "rgl_version"
+
+#ifdef _WIN32
+#define TAPE_HOOK(...) ;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.hpp:51]: No needed. Removed

File: src/api/api.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (prybicki) [reply src/api/api.cpp]: Anyway, I'd put an `#else`, just to please any IDE anyone may use in the future.

**Review by prybicki — APPROVED** [review]
Built & tested, works fine!

**Pull request merged**